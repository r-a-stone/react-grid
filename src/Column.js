import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { StyledRow } from "./Row";

const propTypes = {
  shape: PropTypes.object,
  order: PropTypes.object
};

const defaultProps = {
  shape: {
    xs: 12
  },
  order: {
    xs: 1
  }
};

const StyledColumn = styled("div")`
  box-sizing: border-box;
  &:before,
  &:after {
    box-sizing: border-box;
  }

  padding-right: ${props => `${props.theme.gutters / 2 / 16}rem`};
  padding-left: ${props => `${props.theme.gutters / 2 / 16}rem`};

  ${StyledRow} {
    margin-right: ${props => `${props.theme.gutters / -2 / 16}rem`};
    margin-left: ${props => `${props.theme.gutters / -2 / 16}rem`};
  }

  flex-basis: 100%;
  flex-grow: 0;
  flex-shrink: 1;

  ${props => {
    const { shape, order, breakpoints, columns } = props.theme;
    return breakpoints.reduce((mediaQueries, breakpoint) => {
      const { name, minWidth } = breakpoint;
      if (shape[name] !== undefined || order[name] !== undefined) {
        mediaQueries = `${mediaQueries}
            @media (min-width: ${minWidth / 16}em) {
                ${
                  shape[name] !== undefined
                    ? `flex-basis: ${(shape[name] * 100) / columns}%;`
                    : ""
                }
                ${order[name] !== undefined ? `order: ${order[name]};` : ""}
            }
        `;
      }
      return mediaQueries;
    }, "");
  }}
`;

const Column = ({
  children,
  shape,
  order,
  breakpoints,
  columns,
  gutters,
  ...props
}) => {
  return (
    <StyledColumn
      {...props}
      theme={{ shape, order, breakpoints, columns, gutters }}
    >
      {children}
    </StyledColumn>
  );
};

Column.propTypes = propTypes;
Column.defaultProps = defaultProps;

export default Column;
