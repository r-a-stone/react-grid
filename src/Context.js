import React, { createContext, useState } from "react";
import PropTypes from "prop-types";

const Context = createContext();

const propTypes = {
  columns: PropTypes.number,
  gutters: PropTypes.number,
  breakpoints: PropTypes.array
};

const defaultProps = {
  columns: 12,
  gutters: 24,
  breakpoints: [
    {
      name: "xs",
      minWidth: 0
    },
    {
      name: "sm",
      minWidth: 480
    },
    {
      name: "md",
      minWidth: 768
    },
    {
      name: "lg",
      minWidth: 1024
    }
  ]
};

const GridProvider = ({ children, ...props }) => {
  const [columns, setColumns] = useState(props.columns);
  const [gutters, setGutters] = useState(props.gutters);
  const [breakpoints, setBreakpoints] = useState(props.breakpoints);

  return (
    <Context.Provider
      value={{
        columns,
        gutters,
        breakpoints,

        setColumns,
        setGutters,
        setBreakpoints
      }}
    >
      {children}
    </Context.Provider>
  );
};

GridProvider.propTypes = propTypes;
GridProvider.defaultProps = defaultProps;

export default {
  Provider: GridProvider,
  Consumer: Context.Consumer
};
