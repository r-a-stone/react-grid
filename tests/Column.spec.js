import React from "react";
import renderer from "react-test-renderer";
import Column from "../src/Column";
import Context from "../src/Context";

beforeEach(() => {
  jest.resetModules();
});

test("Column renders without crashing.", () => {
  const component = renderer.create(
    <Column
      columns={12}
      gutters={24}
      breakpoints={[
        {
          name: "xs",
          minWidth: 0
        },
        {
          name: "sm",
          minWidth: 480
        },
        {
          name: "md",
          minWidth: 768
        },
        {
          name: "lg",
          minWidth: 1024
        }
      ]}
    >
      Content
    </Column>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Column with context renders without crashing.", () => {
  const component = renderer.create(
    <Context.Provider>
      <Context.Consumer>
        {({ breakpoints, columns, gutters }) => (
          <Column breakpoints={breakpoints} columns={columns} gutters={gutters}>
            Content
          </Column>
        )}
      </Context.Consumer>
    </Context.Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
