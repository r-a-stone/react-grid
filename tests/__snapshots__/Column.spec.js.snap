// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column renders without crashing. 1`] = `
<div
  className="sc-bwzfXH XRERC"
>
  Content
</div>
`;

exports[`Column with context renders without crashing. 1`] = `
<div
  className="sc-bwzfXH XRERC"
>
  Content
</div>
`;
